simple_network_sim/network_of_populations/__init__.py

```diff
@@ -536,61 +536,69 @@ def getInternalInfectiousContactsInNode(
+        infectious: List[float],
+        susceptibles: Union[int, float],
+        totalInAge: Union[int, float],
+        random_state: np.random.Generator
 ) -> float:
     """
-    From raw contacts (between any two people in different age groups), filters only those contacts that originated from
-    an infectious person and received by a susceptible person. The contact did not yet lead to a new infection, we need
-    to multiply by the infection probability.
-
-    :param contacts: The number of reported contacts
-    :param infectious: The number of infectious people from which the contacts
-                       can originate
-    :param susceptibles: The number of susceptible people to which the contacts
-                         can target
+    Takes a list of contacts (usually coming from other age groups) which should be paired with a list of infectious
+    people from the same group. This function will then estimate the number of infectious contacts those people will
+    produce.
+
+    :param contacts: A list with the number of contacts per incoming group
+    :param infectious: The number of incoming infectious people, paired with the the list of contacts
+    :param susceptibles: The number of susceptible people to which the contacts can target
     :param totalInAge: Total number of people in target population
     :param random_state: Random number generator used for the model
-    :return: The number of potentially infectious contacts.
+    :return: The number of potentially infectious contacts
     """
-    assert isinstance(infectious, int) or infectious.is_integer()
+    if len(contacts) != len(infectious):
+        raise ValueError("contacts and infectious must have the same length")
     assert isinstance(totalInAge, int) or totalInAge.is_integer()
     assert isinstance(susceptibles, int) or susceptibles.is_integer()
 
-    numberOfContacts = random_state.poisson(contacts, size=int(infectious))
-    # If 0, the hypergeometric calls fails, but mathematically is 0
-    numberOfContacts = numberOfContacts[numberOfContacts > 0]
+    inf_max = int(round(max(infectious)))
+    # each np.array represent one original (contact, infectious) pair
+    numbersOfContacts: List[np.array] = []
+    for contact, infected in zip(contacts, infectious):
+        x = random_state.poisson(contact, int(round(infected)))
+        numbersOfContacts.append(np.concatenate([x, np.zeros(inf_max - len(x), dtype=int)]))
     # Number of contacts cannot be higher than the number of people in the node
-    numberOfContacts = np.minimum(numberOfContacts, int(totalInAge))
-    numberOfContacts = random_state.hypergeometric(
-        int(susceptibles),
-        int(totalInAge) - int(susceptibles),
-        numberOfContacts
-    )
-    return cast(float, np.sum(numberOfContacts))
+    numbersOfContacts = np.minimum(np.array(numbersOfContacts).T, int(totalInAge))
+    result = random_state.hypergeometric(
+        [int(susceptibles)] * len(contacts),
```

Comment:
Same as above, why not using `np.full(len(contacts), int(susceptibles)))` rather than the list multiplication?

Reply:
The reason is that I didn't know about `np.full` :P

Reply:
```
timeit.timeit("[312] * 10", globals=globals())
0.10730258299736306
timeit.timeit("np.full(10, 312)", globals=globals())
1.6013900120160542
````
np.full seems slower, though.

Reply:
That's very surprising to me! However, one thing though, is that hypergeom might convert the lists to np arrays internally, in which case:
![image](https://user-images.githubusercontent.com/24376790/88774354-38105b00-d17b-11ea-88dc-4f69ef525f8d.png)

Because I am seeing this:
![image](https://user-images.githubusercontent.com/24376790/88774376-43638680-d17b-11ea-8b42-ba6421d3f2d9.png)

Reply:
Ref: https://stackoverflow.com/questions/5891410/numpy-array-initialization-fill-with-identical-values/44921249#44921249

Reply:
And for some reason that I don't fully understand, `np.zeros(100) + 312` is slightly faster too??